support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -35,22 +37,70 @@ object GiftRecipient {
   val circeDiscriminator = new CirceDiscriminator("giftRecipientType")
 
   private val discriminatorWeekly = "Weekly"
-  private val discriminatorDigiSub = "DigiSub"
+  private val discriminatorDigitalSubscription = "DigitalSubscription"
 
   implicit val e1 = circeDiscriminator.add(deriveEncoder[WeeklyGiftRecipient], discriminatorWeekly)
   implicit val d1 = deriveDecoder[WeeklyGiftRecipient]
-  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubGiftRecipient], discriminatorDigiSub)
-  implicit val d2 = deriveDecoder[DigitalSubGiftRecipient]
+  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubscriptionGiftRecipient], discriminatorDigitalSubscription)
+  implicit val d2 = deriveDecoder[DigitalSubscriptionGiftRecipient]
 
   implicit val encoder: Encoder[GiftRecipient] = Encoder.instance {
     case c: WeeklyGiftRecipient => c.asJson
-    case c: DigitalSubGiftRecipient => c.asJson
+    case c: DigitalSubscriptionGiftRecipient => c.asJson
   }
 
   implicit val decoder: Decoder[GiftRecipient] =
     circeDiscriminator.decode[GiftRecipient](Map(
       discriminatorWeekly -> Decoder[WeeklyGiftRecipient],
-      discriminatorDigiSub -> Decoder[DigitalSubGiftRecipient]
+      discriminatorDigitalSubscription -> Decoder[DigitalSubscriptionGiftRecipient]
+    ))
+}
+
+case class GeneratedGiftCode private(value: String) extends AnyVal
+
+object GeneratedGiftCode {
+
+  def apply(value: String): Option[GeneratedGiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GeneratedGiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GeneratedGiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GeneratedGiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftPurchase {
+  def asDigitalSubscriptionGiftPurchase: Option[DigitalSubscriptionGiftPurchase] =
+    this match { case a: DigitalSubscriptionGiftPurchase => Some(a); case _ => None }
+  def giftRecipient: GiftRecipient
+}
+object GiftPurchase {
+
+  case class WeeklyGiftPurchase(giftRecipient: WeeklyGiftRecipient) extends GiftPurchase
+  case class DigitalSubscriptionGiftPurchase(
+    giftRecipient: DigitalSubscriptionGiftRecipient,
+    giftCode: GeneratedGiftCode,
+  ) extends GiftPurchase
```

Comment:
(re-made comment as it had become "outdated")

new ADT to represent a product specific purchased gift

for DS gifts, so far it also includes the gift code. In the next PR it will also get the last_redemption_date field added.
for GW gifts there is no additional information generated in the subscribe step over a basic GW